### DISCUSSION TITLE
Installing IDAKLU solver on mac

### DISCUSSION THREAD

**Discussion by Thiyaga2025** [pybamm-team/PyBaMM#4258] [Q&A] [answered]
hi all,

sorry for this basic question. I am trying to do degradation modeling and facing convergence issues. One of the suggestions i found in the discussion forum is about using IDAKLU solver. can you please guide me how to install IDAKLU from beginners point of view( i did pip install klu, but that didnt help to overcome the error "ImportError: KLU is not installed"). i did see quite a few threads here but found to be specific. 

saw this link also, https://docs.pybamm.org/en/v0.2.3/install/install-klu.html
should i follow this or is it for developers. 

any help will be really appreciated. thanks.

**Comment by agriyakhetarpal** [comment]
Hi @Thiyaga2025, I can see two things:

1. You're looking at the documentation for a very old version of PyBaMM (v0.2.3) – the latest version is v24.1, and
2. You're trying to install KLU, which isn't available through PyPI and this cannot be `pip`-installed. the IDAKLU solver is available with the PyBaMM wheels itself, and therefore is included in the software you get from PyPI, i.e., when you type `pip install pybamm`.

If you are trying to build from source and want to compile the IDAKLU solver yourself, we provide a comprehensive guide with instructions to follow along for your platform at https://docs.pybamm.org/en/stable/source/user_guide/installation/install-from-source.html. Please feel free to raise an issue or let us know if you find anything that is confusing or broken.

**Reply by Thiyaga2025** [reply]
hi @agriyakhetarpal, thanks a lot for response and sorry for my delayed reply. I was looking for some info and got the old document which i shared. I did pip install pybamm veriosn 24.1. but when i tried to run the code like below,

import pybamm
import matplotlib.pyplot as plt
# 
model = pybamm.lithium_ion.DFN(
    {
        "SEI": "solvent-diffusion limited",
        "SEI porosity change": "true",
        "lithium plating": "partially reversible",
        "lithium plating porosity change": "true",  # alias for "SEI porosity change"
    }
)
param = pybamm.ParameterValues("OKane2022")
solver=pybamm.IDAKLUSolver()

i got error message "ImportError: KLU is not installed". so i tried to pip install KLU. I just want to run the solver from the end user perspective as of now not interested in building from source.
please help me with the commands to run the IDAKLU solver directly with pybamm.
thanks and regards.

**Reply by agriyakhetarpal** [reply, marked as answer]
Thanks, @Thiyaga2025! Could you please let us know your hardware specifications? Are you running an Intel macOS machine or a newer M-series one? If it is the case that you are on the latter, we did not yet have wheels for that architecture in version 24.1, which might explain your problem. You will have to build from source to use the IDAKLU solver. The wheels contain the necessary libraries for the IDAKLU solver, so not having wheels means that we don't officially support the platform, even though you could still use PyBaMM without the IDAKLU solver.

For the time being, you may try out PyBaMM's latest release candidates (version 24.5rc2) with the command `pip install --pre pybamm` and see if that fixes the error for you. We shall release version 24.5 soon (hopefully, this week), which will mark official M-series support through binaries available on PyPI.

**Reply by Thiyaga2025** [reply]
hi @agriyakhetarpal, thanks for the response, the machine is M-series(M1). guess that is what causing the issue. meanwhile will try the other options as you suggested and get back.

best regards

**Reply by agriyakhetarpal** [reply]
PyBaMM v24.5 was released today, which includes official support for Apple silicon platforms. https://pybamm.org/news/2024/05/31/pybamm-24.5/#macos-arm64-m-series-support

**Reply by Thiyaga2025** [reply]
hi @agriyakhetarpal , thanks a lot for the update. will try and update.

**Reply by Thiyaga2025** [reply]
@agriyakhetarpal , 
i upgraded the pybamm version with the command "pip install pybamm --upgrade". when i tried to run the below code,
import pybamm
import matplotlib.pyplot as plt
# 
model = pybamm.lithium_ion.DFN(
    {
        "SEI": "solvent-diffusion limited",
        "SEI porosity change": "true",
        "lithium plating": "partially reversible",
        "lithium plating porosity change": "true",  # alias for "SEI porosity change"
    }
)
parameter_values = pybamm.ParameterValues("OKane2022")
solver=pybamm.IDAKLUSolver()
cycle_number = 100
exp = pybamm.Experiment(
    [
        (
            "Discharge at 1C until 2.75 V (1 second period)",  # ageing cycles
            "Charge at 0.5C until 4.2 V (5 minute period)",
            "Hold at 4.2 V until C/50 (5 minute period)",
        )
    ]
    * cycle_number
)
sim = pybamm.Simulation(model, parameter_values=parameter_values, experiment=exp, solver=solver)
sol = sim.solve()

i got the error.
RuntimeError: .../casadi/core/casadi_os.cpp:166: Assertion "handle!=nullptr" failed:
PluginInterface::load_plugin: Cannot load shared library 'libcasadi_interpolant_bspline.dylib': 
   (
    Searched directories: 1. casadipath from GlobalOptions
                          2. CASADIPATH env var
                          3. PATH env var (Windows)
                          4. LD_LIBRARY_PATH env var (Linux)
                          5. DYLD_LIBRARY_PATH env var (osx)
    A library may be 'not found' even if the file exists:
          * library is not compatible (different compiler[/bitness](http://localhost:8889/bitness))
          * the dependencies are not found
   )
  Tried '' :
    Error code: dlopen(libcasadi_interpolant_bspline.dylib, 0x0005): tried: 'libcasadi_interpolant_bspline.dylib' (no such file), '[/System/Volumes/Preboot/Cryptexes/OSlibcasadi_interpolant_bspline.dylib](http://localhost:8889/System/Volumes/Preboot/Cryptexes/OSlibcasadi_interpolant_bspline.dylib)' (no such file), '[/Users/thiyagarajan/.pyenv/versions/3.12.2/envs/cellmodelling/lib/python3.12/site-packages/pybamm/.dylibs/libcasadi_interpolant_bspline.dylib](http://localhost:8889/Users/thiyagarajan/.pyenv/versions/3.12.2/envs/cellmodelling/lib/python3.12/site-packages/pybamm/.dylibs/libcasadi_interpolant_bspline.dylib)' (no such file), '[/Users/thiyagarajan/.pyenv/versions/3.12.2/lib/libcasadi_interpolant_bspline.dylib](http://localhost:8889/Users/thiyagarajan/.pyenv/versions/3.12.2/lib/libcasadi_interpolant_bspline.dylib)' (no such file), '[/System/Volumes/Preboot/Cryptexes/OS/Users/thiyagarajan/.pyenv/versions/3.12.2/lib/libcasadi_interpolant_bspline.dylib](http://localhost:8889/System/Volumes/Preboot/Cryptexes/OS/Users/thiyagarajan/.pyenv/versions/3.12.2/lib/libcasadi_interpolant_bspline.dylib)' (no such file), '[/opt/homebrew/lib/libcasadi_interpolant_bspline.dylib](http://localhost:8889/opt/homebrew/lib/libcasadi_interpolant_bspline.dylib)' (no such file), '[/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libcasadi_interpolant_bspline.dylib](http://localhost:8889/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libcasadi_interpolant_bspline.dylib)' (no such file), '[/usr/lib/libcasadi_interpolant_bspline.dylib](http://localhost:8889/usr/lib/libcasadi_interpolant_bspline.dylib)' (no such file, not in dyld cache), 'libcasadi_interpolant_bspline.dylib' (no such file), '[/usr/local/lib/libcasadi_interpolant_bspline.dylib](http://localhost:8889/usr/local/lib/libcasadi_interpolant_bspline.dylib)' (no such file), '[/usr/lib/libcasadi_interpolant_bspline.dylib](http://localhost:8889/usr/lib/libcasadi_interpolant_bspline.dylib)' (no such file, not in dyld cache)

 which i found similar to the one in thread https://github.com/pybamm-team/PyBaMM/discussions/4006. how to set CASADIPATH environment variable. can you please help me with this. thanks.

**Reply by agriyakhetarpal** [reply]
Thanks for the test, @Thiyaga2025 – this is a bug, and we faced another similar one in #3783. It's tricky to fix the bug because of how CasADi's plugin system works. In this case, you may set the `CASADIPATH` environment variable to 

`/Users/thiyagarajan/.pyenv/versions/3.12.2/envs/cellmodelling/lib/python3.12/site-packages/casadi/`

and that should most likely work.

**Reply by Thiyaga2025** [reply]
@agriyakhetarpal, thanks for the note. i tried after setting CASADIPATH environment variable path,

in terminal, i typed this command,

export CASADIPATH=${CASADIPATH}:/Users/thiyagarajan/.pyenv/versions/3.12.2/envs/cellmodelling/lib/python3.12/site-packages/casadi/

but still getting the error while running the pybamm simulation. what am i missing here.

**Reply by Thiyaga2025** [reply]
@agriyakhetarpal , it works now, thanks for all the help.
